### PR TITLE
Codec callback interface

### PR DESCRIPTION
**Pull request by raoulsh**

### CODE REVIEW

File: lib/quicly.c [hunk 1]
```diff
@@ -169,6 +146,10 @@ struct st_quicly_conn_t {
      * 0-RTT and 1-RTT context
      */
     struct st_quicly_application_space_t *application;
+    /**
+     * defaulr codec
```
Comment (sknat) [inline lib/quicly.c]: default :)

File: include/quicly.h [hunk 2]
```diff
@@ -210,6 +218,16 @@ typedef struct st_quicly_packet_allocator_t {
     void (*free_packet)(struct st_quicly_packet_allocator_t *self, quicly_datagram_t *packet);
 } quicly_packet_allocator_t;
 
+typedef struct st_quicly_codec_t {
```
Comment (sknat) [inline include/quicly.h]: Maybe `[st_]quicly_crypto_codec` ?

File: include/quicly.h [hunk 3]
```diff
@@ -210,6 +218,16 @@ typedef struct st_quicly_packet_allocator_t {
...
+    size_t (*encrypt_packet)(struct st_quicly_codec_t *self, ptls_aead_context_t *aead, ptls_cipher_context_t *header_protection,
+                             uint8_t *dst, uint8_t *dst_payload_from, uint8_t *first_byte_at, uint64_t packet_number);
+    int (*decrypt_packet)(struct st_quicly_codec_t *self, quicly_conn_t *conn, quicly_pn_space_t *space,
+                             ptls_cipher_context_t *header_protection, ptls_aead_context_t **aead, size_t epoch,
+                             uint64_t *next_expected_pn, quicly_decoded_packet_t *packet);
+    int (*decrypt_packet_callback)(quicly_conn_t *conn, quicly_decoded_packet_t *packet, quicly_pn_space_t *space, size_t epoch, uint64_t pn,
```
Comment (sknat) [inline include/quicly.h]: Just a suggestion, `decrypt_packet_continue` or `decrypt_packet_done` ?

**Review by sknat — COMMENTED** [review]
Cool !
I thinks we might also want to edit cli.c so that it sets the proper crypto_codec callbacks for testing
Also might be worth taking a look at their `make test`

File: include/quicly.h (hunk 3 above)
Comment (AloysAugustin) [reply include/quicly.h]: +1 for decrypt_packet_done

**Review by AloysAugustin — COMMENTED** [review]
If we want to upstream the changes to CMakeLists.txt, we probably want to do that in a separate PR. Apart from that this looks good :)

**Comment by sknat** [comment]
+1 for the separated PR for `CMakelists.txt` might be worth discussing quicly release numbers at the same time.

**Comment by raoulsh** [comment]
> +1 for the separated PR for `CMakelists.txt` might be worth discussing quicly release numbers at the same time.

Yes this commit will not be included in the PR, just for tests with VPP

**Comment by raoulsh** [comment]
> Cool !
> I thinks we might also want to edit cli.c so that it sets the proper crypto_codec callbacks for testing
> Also might be worth taking a look at their `make test`

Tests and cli updated

**Pull request closed without merging**